File: go/bob/bob.go
package bob

import (
	"regexp"
	"strings"
)

// Hey should have a comment documenting it.
func Hey(remark string) string {
	nonWhitespace, _ := regexp.MatchString("[^\t\n\f\r ]", remark)
	if !nonWhitespace {
		return "Fine. Be that way!"
	}

	remark = strings.Trim(remark, " ")
	hasUpper, _ := regexp.MatchString("[A-Z]", remark)
	hasLower, _ := regexp.MatchString("[a-z]", remark)
	isUpper := hasUpper && !hasLower
	isQuestion := remark[len(remark)-1] == '?'

	var resp string
	if isUpper {
		if isQuestion {
			resp = "Calm down, I know what I'm doing!"
		} else {
			resp = "Whoa, chill out!"
		}
	} else {
		if isQuestion {
			resp = "Sure."
		} else {
			resp = "Whatever."
		}
	}
	return resp
}

File: julia/bob/bob.jl
isEmpty(input) = isempty(input) || !occursin(r"\S", input)
isUpper(input) = occursin(r"[A-Z]", input) && input == uppercase(input)
isQuestion(input) = occursin(r"\?\s*$", input)

function bob(stimulus)
    if isEmpty(stimulus)
        return "Fine. Be that way!"
    end
    if isUpper(stimulus)
        if isQuestion(stimulus)
            resp = "Calm down, I know what I'm doing!"
        else
            resp = "Whoa, chill out!"
        end
    else
        if isQuestion(stimulus)
            resp = "Sure."
        else
            resp = "Whatever."
        end
    end
    resp
end